### ISSUE TITLE
Problems with training on multi gpus with horovod

### ISSUE THREAD

**Issue by elv-xuwen** [google/automl#427]
I followed the instructions of training on multi gpus on GCP. My command is:
```
export CUDA_DEVICE_ORDER='PCI_BUS_ID'
export CUDA_VISIBLE_DEVICES='0,1'
horovodrun -np 2 -H localhost:2 python main.py --mode=train_and_eval \
    --training_file_pattern=train_tfrecord/*.tfrecord \
    --validation_file_pattern=val_tfrecord/*.tfrecord \
    --model_name=efficientdet-d2 \
    --model_dir=/tmp/efficientdet-d2-coco  \
    --backbone_ckpt=efficientnet-b2 \
    --train_batch_size=8 \
    --eval_batch_size=8 --eval_samples=5000 \
    --num_epochs=10  \
    --hparams="use_horovod=True" \
    --use_tpu=False
```

And I met two issues:
1. I set number to be 2, I supposed it would run one process on 2 gpus, but it runs two processes on both gpus.
Like this:
![image](https://user-images.githubusercontent.com/40640724/82387427-da79c880-99eb-11ea-8b68-c7a319a6b393.png)
2. The training gets hung after outputting logs:

_[1,0]<stderr>:2020-05-19 23:00:45.870715: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
[1,0]<stderr>:2020-05-19 23:00:45.870726: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
[1,0]<stderr>:Skipping registering GPU devices...
[1,0]<stderr>:2020-05-19 23:00:45.870777: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
[1,0]<stderr>:2020-05-19 23:00:45.870794: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      0 1
[1,0]<stderr>:2020-05-19 23:00:45.870802: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 0:   N Y
[1,0]<stderr>:2020-05-19 23:00:45.870808: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 1:   Y N
[1,1]<stderr>:2020-05-19 23:00:45.871034: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
[1,1]<stderr>:2020-05-19 23:00:45.872130: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties:
[1,1]<stderr>:pciBusID: 0000:00:04.0 name: Tesla P100-PCIE-16GB computeCapability: 6.0
[1,1]<stderr>:coreClock: 1.3285GHz coreCount: 56 deviceMemorySize: 15.90GiB deviceMemoryBandwidth: 681.88GiB/s
[1,1]<stderr>:2020-05-19 23:00:45.872261: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
[1,1]<stderr>:2020-05-19 23:00:45.872857: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 1 with properties:
[1,1]<stderr>:pciBusID: 0000:00:05.0 name: Tesla P100-PCIE-16GB computeCapability: 6.0
[1,1]<stderr>:coreClock: 1.3285GHz coreCount: 56 deviceMemorySize: 15.90GiB deviceMemoryBandwidth: 681.88GiB/s
[1,1]<stderr>:2020-05-19 23:00:45.873051: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
[1,1]<stderr>:2020-05-19 23:00:45.873178: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcublas.so.10'; dlerror: libcublas.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
[1,1]<stderr>:2020-05-19 23:00:45.873297: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcufft.so.10'; dlerror: libcufft.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
[1,1]<stderr>:2020-05-19 23:00:45.873414: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcurand.so.10'; dlerror: libcurand.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
[1,1]<stderr>:2020-05-19 23:00:45.873552: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusolver.so.10'; dlerror: libcusolver.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
[1,1]<stderr>:2020-05-19 23:00:45.873679: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusparse.so.10'; dlerror: libcusparse.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
[1,1]<stderr>:2020-05-19 23:00:45.873748: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
[1,1]<stderr>:2020-05-19 23:00:45.873761: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
[1,1]<stderr>:Skipping registering GPU devices...
[1,1]<stderr>:2020-05-19 23:00:45.873823: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
[1,1]<stderr>:2020-05-19 23:00:45.873847: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      0 1
[1,1]<stderr>:2020-05-19 23:00:45.873859: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 0:   N Y
[1,1]<stderr>:2020-05-19 23:00:45.873866: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 1:   Y N
[1,1]<stderr>:WARNING:tensorflow:From /home/elv-xuwen/elv-ml/object-detection/automl/efficientdet/venv_efficientdet_offi/lib/python3.7/site-packages/tensorflow/python/training/saver.py:1077: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
[1,1]<stderr>:Instructions for updating:
[1,1]<stderr>:Use standard file utilities to get mtimes.
[1,1]<stderr>:W0519 23:01:20.766196 139833721849216 deprecation.py:323] From /home/elv-xuwen/elv-ml/object-detection/automl/efficientdet/venv_efficientdet_offi/lib/python3.7/site-packages/tensorflow/python/training/saver.py:1077: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
[1,1]<stderr>:Instructions for updating:
[1,1]<stderr>:Use standard file utilities to get mtimes.
[1,0]<stderr>:WARNING:tensorflow:From /home/elv-xuwen/elv-ml/object-detection/automl/efficientdet/venv_efficientdet_offi/lib/python3.7/site-packages/tensorflow/python/training/saver.py:1077: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
[1,0]<stderr>:Instructions for updating:
[1,0]<stderr>:Use standard file utilities to get mtimes.
[1,0]<stderr>:W0519 23:01:21.952915 140617459418496 deprecation.py:323] From /home/elv-xuwen/elv-ml/object-detection/automl/efficientdet/venv_efficientdet_offi/lib/python3.7/site-packages/tensorflow/python/training/saver.py:1077: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
[1,0]<stderr>:Instructions for updating:
[1,0]<stderr>:Use standard file utilities to get mtimes._

**Comment by fsx950223** [comment]
It's training.

**Comment by elv-xuwen** [comment]
> It's training and remove your exports

Hi @fsx950223  Thanks for your reply! 
If it's training, why is the GPU usage that low? And do you have any idea about the first problem?

**Comment by fsx950223** [comment]
> > It's training and remove your exports
> 
> Hi @fsx950223 Thanks for your reply!
> If it's training, why is the GPU usage that low? And do you have any idea about the first problem?

Could you check your gpu environment?

**Comment by xiaoxin05** [comment]
Hello, I am going to use dual GPU to run this task. Have you solved the above problems

**Comment by xiaoxin05** [comment]
Hello, I am going to use dual GPU to run this task. Have you solved the above problems @elv-xuwen

**Comment by romainvo** [comment]
Hi, first thank you very much for your work, it's awesome !
I tried to train efficientdet-d3 on my custom dataset with multi-gpu. It trains fine in single-gpu but it looks like the multi-gpu implementation is a bit more complext han what you did in #413 . I got this error after one epoch : 
```
[1,0]<stderr>:[2020-07-10 11:14:01.626480: W horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
[1,0]<stderr>:Stalled ranks:
[1,0]<stderr>:0: [HorovodBroadcast_box_net_box_0_bias_0, HorovodBroadcast_box_net_box_0_bias_Momentum_0, HorovodBroadcast_box_net_box_0_bn_3_beta_0, HorovodBroadcast_box_net_box_0_bn_3_beta_Momentum_0, HorovodBroadcast_box_net_box_0_bn_3_gamma_0, HorovodBroadcast_box_net_box_0_bn_3_gamma_Momentum_0 ...]
[1,0]<stderr>:1: [DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1000_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1002_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1006_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1011_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1015_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1016_0 ...]
[1,0]<stderr>:2: [DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1000_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1002_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1006_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1011_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1015_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1016_0 ...]
[1,0]<stderr>:3: [DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1000_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1002_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1006_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1011_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1015_0, DistributedMomentumOptimizer_Allreduce/HorovodAllreduce_gradients_AddN_1016_0 ...][1,0]<stderr>:
```
I think it's related to the broadcast of restored variables to all gpus, it needs to be done only on rank0 gpu ? Is it investigated by @fsx950223 or @mingxingtan  right now ? 
It also looks like the warmup init step and other lr scheduling step are not scaled with hvd.size().  I modified this in det_model_fn.py : 
``` 
  # Set up training loss and learning rate.
  update_learning_rate_schedule_parameters(params)

  ###############################################################
  #Custom modification 
  b_s = (
    params['batch_size'] * params['num_shards']
    if params['strategy'] == 'tpu' else params['batch_size'])
  params['lr_warmup_init'] = (
    params['lr_warmup_init'] * b_s / _DEFAULT_BATCH_SIZE)
  if params['strategy'] == 'horovod':  
    import horovod.tensorflow as hvd
    params['lr_warmup_init'] = params['lr_warmup_init'] * hvd.size()
    params['learning_rate'] = params['learning_rate'] * hvd_size()
    params['lr_warmup_step'] = int(params['lr_warmup_step'] / hvd.size())
    params['first_lr_drop_step'] = int(params['first_lr_drop_step'] / hvd.size())
    params['second_lr_drop_step'] = int(params['second_lr_drop_step'] / hvd.size())
    params['total_steps'] = int(params['total_steps'] / hvd.size())
  ###############################################################
  ###############################################################

  global_step = tf.train.get_or_create_global_step()
  learning_rate = learning_rate_schedule(params, global_step)
``` 

It feels like the multi-gpu messed a lot with the training loop also... : 
![Capture d’écran du 2020-07-10 14-21-25](https://user-images.githubusercontent.com/43675867/87153832-ff820d80-c2af-11ea-8e3a-3d18a7efc092.png)

**Comment by drwaltman** [comment]
Hi @fsx950223 @mingxingtan, I am seeing the same issues as @romainvo for training with multiple GPUs via Horovod.  As far as I can tell, it looks like the following steps from the [Horovod TensorFlow docs](https://horovod.readthedocs.io/en/stable/tensorflow.html) still need to be implemented:

> 3. Scale the learning rate by the number of workers.
Effective batch size in synchronous distributed training is scaled by the number of workers. An increase in learning rate compensates for the increased batch size.

* It looks like @romainvo has addressed step 3 with his modifications above.

> 5. Broadcast the initial variable states from rank 0 to all other processes.
This is necessary to ensure consistent initialization of all workers when training is started with random weights or restored from a checkpoint.
For **TensorFlow v1**, add hvd.BroadcastGlobalVariablesHook(0) when using a MonitoredTrainingSession. When not using MonitoredTrainingSession, execute the hvd.broadcast_global_variables op after global variables have been initialized.
For **TensorFlow v2**, use hvd.broadcast_variables after models and optimizers have been initialized.

* The stalled ranks issue mentioned in https://github.com/google/automl/issues/563 seems related to step 5 since it occurs after the first eval has finished.  I noticed `hvd.BroadcastGlobalVariablesHook(0)` is used in [det_model_fn.py](https://github.com/google/automl/blob/master/efficientdet/det_model_fn.py#L397), but I can't find use of `MonitoredTrainingSession`.  Should `hvd.broadcast_global_variables` or `hvd.broadcast_variables` be used instead as documented?

> 6. Modify your code to save checkpoints only on worker 0 to prevent other workers from corrupting them.
For **TensorFlow v1**, accomplish this by passing checkpoint_dir=None to tf.train.MonitoredTrainingSession if hvd.rank() != 0.
For **TensorFlow v2**, construct a tf.train.Checkpoint and only call checkpoint.save() when hvd.rank() == 0.

* It seems like step 6 should already be covered by this [line](https://github.com/google/automl/blob/master/efficientdet/main.py#L237), but I am seeing overlapping data points from a single training run:

    * <img width="1792" alt="Screen Shot 2020-08-02 at 2 58 57 PM" src="https://user-images.githubusercontent.com/556045/89133610-24922680-d4d2-11ea-9c09-6db031ea7d3d.png">
